File: App/General/Layout/MBSceneStackView.m

#import "MBSceneStackView.h"
#import <RFKit/NSArray+RFKit.h>

@interface MBSceneStackView ()
@property NSInteger activeSceneIndex;
@end

@implementation MBSceneStackView

- (void)setScenes:(NSArray<NSArray<UIView *> *> *)scenes {
    _scenes = scenes;
    [scenes enumerateObjectsUsingBlock:^(NSArray<UIView *> *vs, NSUInteger idx, BOOL * _Nonnull stop) {
        for (UIView *v in vs) {
            v.hidden = (idx != self.activeSceneIndex);
        }
    }];
}

- (void)setActiveSceneWithIndex:(NSInteger)index animated:(BOOL)animated {
    NSArray<NSArray<UIView *> *> *scenes = self.scenes;
    NSInteger count = scenes.count;
    if (!count || index >= count) return;
    
    NSArray *scNeedHide = [scenes rf_objectAtIndex:self.activeSceneIndex];
    NSArray *scNeedShow = [scenes rf_objectAtIndex:index];
    CGFloat transform = (self.activeSceneIndex < index)? 100 : -100;
    self.activeSceneIndex = index;
    if (self.onSceneChanged) {
        self.onSceneChanged(self, index);
    }
    if (!animated) {
        for (UIView *v in scNeedHide) {
            v.hidden = YES;
        }
        for (UIView *v in scNeedShow) {
            v.hidden = NO;
        }
        return;
    }
    
    [UIView animateWithDuration:.2 delay:0 options:UIViewAnimationOptionCurveEaseIn animations:^{
        for (UIView *v in scNeedHide) {
            v.alpha = 0;
            v.transform = CGAffineTransformMakeTranslation(-transform, 0);
        }
    } completion:^(BOOL finished) {
        for (UIView *v in scNeedHide) {
            v.hidden = YES;
            v.alpha = 1;
            v.transform = CGAffineTransformIdentity;
        }
        for (UIView *v in scNeedShow) {
            v.hidden = NO;
            v.alpha = 0;
            v.transform = CGAffineTransformMakeTranslation(transform, 0);
        }
        [self layoutIfNeeded];
        
        [UIView animateWithDuration:.2 delay:0 options:UIViewAnimationOptionCurveEaseOut animations:^{
            for (UIView *v in scNeedShow) {
                v.alpha = 1;
                v.transform = CGAffineTransformIdentity;
            }
        } completion:nil];
    }];
}

- (void)nextSceneAnimated:(BOOL)animated {
    [self setActiveSceneWithIndex:self.activeSceneIndex + 1 animated:animated];
}

- (void)previousSceneAnimated:(BOOL)animated {
    [self setActiveSceneWithIndex:self.activeSceneIndex - 1 animated:animated];
}

@end

File: App/General/Layout/MBSceneStackView.swift
/*
 MBSceneStackView

 Copyright © 2018, 2023 BB9z.
 https://github.com/BB9z/iOS-Project-Template

 The MIT License
 https://opensource.org/licenses/MIT
 */

import UIKit

// @MBDependency:2
/**
 基于 UIStackView 的界面切换

 把 UIStackView 中的 view 分成若干组，每次显示一组，并支持组间切换显示

 UIStackView based views switching

 Divide the views in the UIStackView into several groups, display one group at a time, and support switching between groups
 */
class MBSceneStackView: UIStackView {

    private(set) var activeSceneIndex = 0

    var scenes = [[UIView]]() {
        didSet {
            for (index, views) in scenes.enumerated() {
                views.forEach { $0.isHidden = index != activeSceneIndex }
            }
        }
    }

    var onSceneChanged: ((MBSceneStackView, Int) -> Void)?

    func setActiveScene(at index: Int, animated: Bool, layoutView: UIView? = nil) {
        guard index < scenes.count else { return }

        let sceneNeedHide = scenes[activeSceneIndex]
        let sceneNeedShow = scenes[index]
        let transform: CGFloat = (activeSceneIndex < index) ? 100 : -100
        activeSceneIndex = index

        if let onSceneChanged = onSceneChanged {
            onSceneChanged(self, index)
        }

        if !animated {
            sceneNeedHide.forEach { $0.isHidden = true }
            sceneNeedShow.forEach { $0.isHidden = false }
            return
        }

        UIView.animate(withDuration: 0.2, delay: 0, options: .curveEaseIn) {
            sceneNeedHide.forEach {
                $0.alpha = 0
                $0.transform = CGAffineTransform(translationX: -transform, y: 0)
            }
        } completion: { _ in
            sceneNeedHide.forEach {
                $0.isHidden = true
                $0.alpha = 1
                $0.transform = CGAffineTransform.identity
            }
            sceneNeedShow.forEach {
                $0.isHidden = false
                $0.alpha = 0
                $0.transform = CGAffineTransform(translationX: transform, y: 0)
            }
            (layoutView ?? self).layoutIfNeeded()

            UIView.animate(withDuration: 0.2, delay: 0, options: .curveEaseOut) {
                sceneNeedShow.forEach {
                    $0.alpha = 1
                    $0.transform = CGAffineTransform.identity
                }
            }
        }
    }

    func nextScene(animated: Bool) {
        setActiveScene(at: activeSceneIndex + 1, animated: animated)
    }

    func previousScene(animated: Bool) {
        setActiveScene(at: activeSceneIndex - 1, animated: animated)
    }
}
